File: lintcode/cpp/3_Sum_Closest.cpp
class Solution {
public:    
    /**
     * @param numbers: Give an array numbers of n integer
     * @param target: An integer
     * @return: return the sum of the three integers, the sum closest target.
     */
    int threeSumClosest(vector<int> nums, int target) {
        // write your code here
        if(nums.size() < 3) throw runtime_error("wrong input");
        
        sort(nums.begin(), nums.end());
        
        int n = nums.size();
        int minDis = INT_MAX;
        int result;
        
        for(int i = 0; i < n-2; i++){
            int j = i+1;
            int k = n-1;
            
            while(j < k){
                int sum = nums[i] + nums[j] + nums[k];
                
                // for each candidate, find the closest one
                int dis = abs(sum - target);
                if(dis < minDis) {
                    minDis = dis;
                    result = sum;
                }
                
                if(sum < target) j++;
                else k--;
            }
        }
        
        return result;
    }
};

File: lintcode/java/3SumClosest.java
public class Solution {
    /**
     * @param numbers: Give an array numbers of n integer
     * @param target : An integer
     * @return : return the sum of the three integers, the sum closest target.
     */
    public int threeSumClosest(int[] numbers, int target) {
        if (numbers.length < 3) throw new IllegalArgumentException("xxx");

        Arrays.sort(numbers);
        
        int n = numbers.length;
        int minDis = Integer.MAX_VALUE;
        int minSum = 0;
        for (int i = 0; i < n; i++) {
            int j = i + 1;
            int k = n - 1;
            while (j < k) {
                int t = numbers[i] + numbers[j] + numbers[k];
                int dis = Math.abs(t-target);
                if(dis < minDis){
                    minSum = t;
                    minDis = dis;
                }
                if(t > target) k--;
                else if (t == target) return target;
                else j++;
            }
        }
        return minSum;
    }
}
